airlock/api.py

```diff
@@ -315,12 +336,21 @@ def set_status(
         release_request.__dict__["status"] = to_status
 
     def add_file_to_request(
-        self, release_request: ReleaseRequest, relpath: Path, user: User
-    ):
+        self,
+        release_request: ReleaseRequest,
+        relpath: Path,
+        user: User,
+        group_name: Optional[str] = "default",
+    ) -> ReleaseRequest:
         """Add a file to a request.
 
         Subclasses should call super().add_file_to_request(...) to do the
-        copying, then record the file metadata as needed.
+        copying, then create/retrieve the file group and record the file
+        metadata as needed.
+
+        After calling add_file_to_request, the current release_request's
+        filegroups will be out of date. Subclasses should return a new
+        ReleaseRequest to ensure filegroups are current.
```

Comment:
I'm not sure about this. I know the `set_state` method changes the ReleaseRequest object, but that doesn't feel like a thing that should be done unless it has to.  I've chosen to make `add_file_to_request` return the release_request here, and a new one in the provider api, but the provider api could modify the existing release_request. In the views it doesn't actually need to, because adding a request redirects to the request file view, where the request will be instantiated again and will have the updated groups.

Reply:
Yeah, I agree this is an awkward case. I originally wrote `set_status()` to return a new instance in a similar manner for similar reasons.  But the old instance can still be used, so feels like a foot gun (and indeed it did bite me). Unlike Closure, Python programmers are not used to immutable data types, except for, famously, strings. But if we document this aspect of the API, its probably ok?

And like you say, its unlikely to matter in the current usage, but in future that might change.

Also, lists are mutable members of frozen dataclasses, so you *can* actually mutate them as normal, rather than having to assign to __dict__ as I did with `.status`. As you've added a list, ReleaseRequest is no longer hashable, but I but assume that's not been an issue. I don't its avoidable anyway.

In that case, all `frozen=True` does is prevent mutation of normal attributes.

So I actually think mutating the list is probably fine?  But probably worth discussing with rest of RAP (separate to this PR)?

Whatever we decide, there's probably value in doing the same thing across the board for all the APIs that mutate - either mutate passed object, or return a new object.  If we do return new, it would be nice to have some way to force the caller to not use that instance again (aka tainted objects).

Reply:
OK, I've left it as is (creating the new object). Something to discuss at the next code walkthrough, maybe. Do we want ReleaseRequest to be hashable? It's not been an issue yet, but a jobserver provider API might want it? I'm going to leave that for later too, though.